classes/profile.php

```diff
@@ -193,12 +193,32 @@ public function save_record(): int {
         $this->raw_set('timemodified', $now);
         $this->raw_set('usermodified', $USER->id);
 
+        $storeduser = (int) $this->raw_get('userid');
+
         // We may not have obtained a valid userid when the profile record was created.
-        // If the current profile userid is 0, but there's a valid $USER->id, update it.
-        if ($USER->id && ! $this->raw_get('userid')) {
+        // If the stored userid is 0, and there's now a valid $USER->id, update the stored userid.
+        if ($USER->id && !$storeduser) {
+            $this->raw_set('userid', $USER->id);
+        }
+
+        // If current userid isn't guest, but stored user is guest, update it.
+        $guestrole = get_guest_role();
```

Comment:
I also don't think this is the correct check. This returns the roleid for guest users, and I think it the value to check here might need to be `$CFG->siteguest` or similar instead